0.10.0-nightly-8004f11 crashing multiple times a day!

I am accepting traffic via http and also udp. The engine is tsm1 , a single server and on a completely clean install. It always "appears" to crash right after a query but the same query can be run again after the restart and work fine. Here are snippets of some crashes from a few days ago:

```
[http] 2016/01/05 15:42:03 10.25.2.164 - - [05/Jan/2016:15:42:03 +0000] POST /write?db=myTestDB&precision=s&u=&p= HTTP/1.1 204 0 - EventMachine HttpClient de31571d-b3c2-11e5-9ccf-000000000000 1.847503ms
[query] 2016/01/05 15:42:05 SELECT recBytes FROM elb."default".elbs WHERE elb =~ /elb-api-server$/ AND time > now() - 1h
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x612a56]

goroutine 1743764 [running]:
github.com/influxdb/influxdb/tsdb.(*RawExecutor).execute(0xc82b17c8a0, 0xc83d5bd3e0, 0xc83d5bd260)
        /tmp/tmp.eqCMvBE7Hx/src/github.com/influxdb/influxdb/tsdb/raw.go:229 +0xef6
created by github.com/influxdb/influxdb/tsdb.(*RawExecutor).Execute
        /tmp/tmp.eqCMvBE7Hx/src/github.com/influxdb/influxdb/tsdb/raw.go:59 +0x67
```

```
[http] 2016/01/05 17:45:07 10.25.2.164 - - [05/Jan/2016:17:45:07 +0000] POST /write?db=myTestDB&precision=s&u=&p= HTTP/1.1 204 0 - EventMachine HttpClient 0f0e0c73-b3d4-11e5-ac3f-000000000000 1.853793ms
[query] 2016/01/05 17:45:07 SELECT count(backendTime) FROM elb."default".elbs WHERE elb =~ /elb-api-server$/ AND (elbStatusCode = '500' OR elbStatusCode = '504') AND time > now() - 30m GROUP BY time(1s)
[query] 2016/01/05 17:45:07 SELECT recBytes FROM elb."default".elbs WHERE elb =~ /elb-api-server$/ AND time > now() - 30m
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x612a56]

goroutine 181186 [running]:
github.com/influxdb/influxdb/tsdb.(*RawExecutor).execute(0xc827bdb890, 0xc829caafc0, 0xc829caae40)
        /tmp/tmp.eqCMvBE7Hx/src/github.com/influxdb/influxdb/tsdb/raw.go:229 +0xef6
created by github.com/influxdb/influxdb/tsdb.(*RawExecutor).Execute
        /tmp/tmp.eqCMvBE7Hx/src/github.com/influxdb/influxdb/tsdb/raw.go:59 +0x67
```

```
[http] 2016/01/05 20:33:02 10.25.2.164 - - [05/Jan/2016:20:33:02 +0000] POST /write?db=myTestDB&precision=s&u=&p= HTTP/1.1 204 0 - EventMachine HttpClient 84934e4e-b3eb-11e5-9b14-000000000000 2.23985ms
[query] 2016/01/05 20:33:08 SELECT recBytes FROM elb."default".elbs WHERE elb =~ /elb-api-server$/ AND time > now() - 30m
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x612a56]

goroutine 364592 [running]:
github.com/influxdb/influxdb/tsdb.(*RawExecutor).execute(0xc833450a80, 0xc825c06360, 0xc825c06180)
        /tmp/tmp.eqCMvBE7Hx/src/github.com/influxdb/influxdb/tsdb/raw.go:229 +0xef6
created by github.com/influxdb/influxdb/tsdb.(*RawExecutor).Execute
        /tmp/tmp.eqCMvBE7Hx/src/github.com/influxdb/influxdb/tsdb/raw.go:59 +0x67
```

```
[http] 2016/01/05 21:03:00 10.25.2.164 - - [05/Jan/2016:21:03:00 +0000] POST /write?db=myTestDB&precision=s&u=&p= HTTP/1.1 204 0 - EventMachine HttpClient b4663e99-b3ef-11e5-8d48-000000000000 2.154557ms
[query] 2016/01/05 21:03:08 SELECT recBytes FROM elb."default".elbs WHERE elb =~ /elb-api-server$/ AND time > now() - 30m
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x612a56]

goroutine 10318 [running]:
github.com/influxdb/influxdb/tsdb.(*RawExecutor).execute(0xc823da2d20, 0xc82a736540, 0xc82a7363c0)
        /tmp/tmp.eqCMvBE7Hx/src/github.com/influxdb/influxdb/tsdb/raw.go:229 +0xef6
created by github.com/influxdb/influxdb/tsdb.(*RawExecutor).Execute
        /tmp/tmp.eqCMvBE7Hx/src/github.com/influxdb/influxdb/tsdb/raw.go:59 +0x67
```

```
[http] 2016/01/05 21:15:04 10.25.2.164 - - [05/Jan/2016:21:15:04 +0000] POST /write?db=myTestDB&precision=s&u=&p= HTTP/1.1 204 0 - EventMachine HttpClient 638db845-b3f1-11e5-98b0-000000000000 2.1595ms
[query] 2016/01/05 21:15:09 SELECT recBytes FROM elb."default".elbs WHERE elb =~ /elb-api-server$/ AND time > now() - 30m
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x612a56]

goroutine 19107 [running]:
github.com/influxdb/influxdb/tsdb.(*RawExecutor).execute(0xc829bdfb90, 0xc8244c0780, 0xc8244c0600)
        /tmp/tmp.eqCMvBE7Hx/src/github.com/influxdb/influxdb/tsdb/raw.go:229 +0xef6
created by github.com/influxdb/influxdb/tsdb.(*RawExecutor).Execute
        /tmp/tmp.eqCMvBE7Hx/src/github.com/influxdb/influxdb/tsdb/raw.go:59 +0x67
```

```
[http] 2016/01/05 22:33:35 10.25.2.164 - - [05/Jan/2016:22:33:35 +0000] POST /write?db=myTestDB&precision=s&u=&p= HTTP/1.1 204 0 - EventMachine HttpClient 5b8b79d3-b3fc-11e5-8cf1-000000000000 1.777596ms
[query] 2016/01/05 22:33:40 SELECT count(backendTime) FROM elb."default".elbs WHERE elb =~ /elb-api-server$/ AND (elbStatusCode = '500' OR elbStatusCode = '504') AND time > now() - 30m GROUP BY time(1s)
[query] 2016/01/05 22:33:40 SELECT recBytes FROM elb."default".elbs WHERE elb =~ /elb-api-server$/ AND time > now() - 30m
[query] 2016/01/05 22:33:40 SELECT count(backendTime) FROM elb."default".elbs WHERE elb =~ /elb-api-server$/ AND elbStatusCode != '200' AND elbStatusCode != '201' AND elbStatusCode != '204' AND time > now() - 30m GROUP BY time(1s), elbStatusCode
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x612a56]

goroutine 108404 [running]:
github.com/influxdb/influxdb/tsdb.(*RawExecutor).execute(0xc82116d9b0, 0xc827ce4600, 0xc827ce4420)
        /tmp/tmp.eqCMvBE7Hx/src/github.com/influxdb/influxdb/tsdb/raw.go:229 +0xef6
created by github.com/influxdb/influxdb/tsdb.(*RawExecutor).Execute
        /tmp/tmp.eqCMvBE7Hx/src/github.com/influxdb/influxdb/tsdb/raw.go:59 +0x67
```
If there is some kind of tracing/debugging switch I can turn on to help debug this, let know.

John Hayden

I've seen this as well:

```
[query] SELECT * FROM database."default".measurement WHERE time > 403316h AND time < 403484h AND tag = 'value' AND tag2 = 'value2' GROUP BY tag3 LIMIT 1
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x633045]

goroutine 9915771 [running]:
github.com/influxdb/influxdb/tsdb.(*RawExecutor).execute(0xc2432dc6f0, 0xc2384aef00, 0xc2384aee40)
        /tmp/tmp.DxFelBPhVV/src/github.com/influxdb/influxdb/tsdb/raw.go:229 +0xf55
created by github.com/influxdb/influxdb/tsdb.(*RawExecutor).Execute
        /tmp/tmp.DxFelBPhVV/src/github.com/influxdb/influxdb/tsdb/raw.go:59 +0x64
```

I know how to recreate this error! Due to a crash in my testDB and the fact I was troubleshooting a different issue for 4 days, I had a db with a gap of 4 days of no data. When I restarted the db, it crashed as soon as a query came in from grafana.

[query] 2016/01/27 23:21:53 SELECT recBytes FROM elb."default".elbs WHERE elb =~ /elb-api-server$/ AND time > now() - 1d
[query] 2016/01/27 23:21:53 SELECT count(backendTime) FROM elb."default".elbs WHERE elb =~ /elb-api-server$/ AND elbStatusCode != '200' AND elbStatusCode != '201' AND elbStatusCode != '204' AND time > now() - 1d GROUP BY time(1m), elbStatusCode
[query] 2016/01/27 23:21:53 SELECT count(backendTime) FROM elb."default".elbs WHERE elb =~ /elb-api-server$/ AND (elbStatusCode = '500' OR elbStatusCode = '504') AND time > now() - 1d GROUP BY time(1m)
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x612a56]

goroutine 81 [running]:
github.com/influxdb/influxdb/tsdb.(*RawExecutor).execute(0xc82a1d8f90, 0xc823124a20, 0xc8231248a0)
        /tmp/tmp.eqCMvBE7Hx/src/github.com/influxdb/influxdb/tsdb/raw.go:229 +0xef6
created by github.com/influxdb/influxdb/tsdb.(*RawExecutor).Execute
        /tmp/tmp.eqCMvBE7Hx/src/github.com/influxdb/influxdb/tsdb/raw.go:59 +0x67

So it appears that if there is no data (or it can't find any within parameters), it crashes. I hope this helps.

Question: Is it worth upgrading to the latest nightly build?

go build -o /tmp/influxd ./cmd/influxd
cmd/influxd/run/command.go:16:2: cannot find package "
github.com/BurntSushi/toml" in any of:
/home/ec2-user/.gvm/gos/go1.4/src/github.com/BurntSushi/toml (from $GOROOT)
/tmp/influxbuild/src/github.com/BurntSushi/toml (from $GOPATH)

John Hayden

On Wed, Feb 17, 2016 at 3:07 PM, Jonathan A. Sternberg <
notifications@github.com> wrote:

> If you have go installed, you should be able to just do this in a new
> directory.
>
> $ mkdir /tmp/influxbuild && cd /tmp/influxbuild
> $ export GOPATH=/tmp/influxbuild
> $ mkdir -p src/github.com/influxdb
> $ git clone git://github.com/influxdata/influxdb -b js-5303-mapped-chunks-nil-panic src/github.com/influxdb/influxdb
> $ cd src/github.com/influxdb/influxdb
> $ go get -u ./...
> $ go build -o /tmp/influxd ./cmd/influxd
>
> Then you can just run the binary at /tmp/influxd.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/influxdata/influxdb/issues/5303#issuecomment-185380919>
> .
>

--- a/tsdb/raw.go
+++ b/tsdb/raw.go
@@ -224,6 +224,11 @@ func (e *RawExecutor) execute(out chan *models.Row, closing <-chan struct{}) {
 			}
 		}
 
+		// Protect against none of the mappers producing any output.
+		if chunkedOutput == nil {
+			continue
+		}
+
 		if ascending {
 			// Sort the values by time first so we can then handle offset and limit
 			sort.Sort(MapperValues(chunkedOutput.Values))

